### PR TITLE
Add Logspout functionality for out-of-the-box Docker logging

### PR DESCRIPTION
**Pull request by dm**
(Rebased and squashed changes from #110)

Add direct link to baked configurations, for debugging purposes.
Separate Logspout service into its own `yml`
Standardize Markdown formatting for program names
Do not auto-create indexes
Remove comments & UDP input

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -49,7 +52,14 @@ You can also choose to run it in background (detached mode):
...
+Now that the stack is running, you'll want to inject logs in it. The shipped Logstash configuration allows you to send content via tcp:
+
+If you want to include a running Logspout container, ensure the additional
+`logspout.yml` file is included in the parameters:
+
+```bash
+$ docker-compose -f docker-compose.yml -f logspout.yml up
+```
 
 ```bash
```
Comment (antoineco) [inline README.md]: I think you wanted to keep that command above the injected text.

File: README.md [hunk 2]
```diff
@@ -59,6 +69,8 @@ And then access Kibana UI by hitting [http://localhost:5601](http://localhost:56
 
 *NOTE*: You'll need to inject data into Logstash before being able to configure a Logstash index pattern in Kibana. Then all you should have to do is to hit the create button.
 
+By default, only the `logspout-*` index is automatically created to provide out-of-the-box indexing of Docker Log messages.
```
Comment (antoineco) [inline README.md]: Why `logspout-*` and not the default `logstash-*` pattern?

File: README.md [hunk 3]
```diff
@@ -67,9 +79,29 @@ By default, the stack exposes the following ports:
 * 9300: Elasticsearch TCP transport
 * 5601: Kibana
 
-*WARNING*: If you're using *boot2docker*, you must access it via the *boot2docker* IP address instead of *localhost*.
+*WARNING*: If you're using `boot2docker`, you must access it via the `boot2docker` IP address instead of `localhost`.
 
-*WARNING*: If you're using *Docker Toolbox*, you must access it via the *docker-machine* IP address instead of *localhost*.
+*WARNING*: If you're using *Docker Toolbox*, you must access it via the `docker-machine` IP address instead of `localhost`.
+
+# Create Indexes
```
Comment (antoineco) [inline README.md]: Index ***pattern***

File: README.md [hunk 4]
```diff
@@ -81,6 +113,28 @@ The Kibana default configuration is stored in `kibana/config/kibana.yml`.
 
 It is also possible to map the entire `config` directory instead of a single file.
 
+## How can I tune Logstash configuration?
```
Comment (antoineco) [inline README.md]: Duplicate of the paragraph below.

File: elasticsearch/config/.placeholder [hunk 5]
```diff
@@ -0,0 +1,5 @@
+Ensure the existence of the parent folder.
```
Comment (antoineco) [inline elasticsearch/config/.placeholder]: File not needed anymore.

File: kibana/config/.placeholder [hunk 6]
```diff
@@ -0,0 +1,5 @@
+Ensure the existence of the parent folder.
```
Comment (antoineco) [inline kibana/config/.placeholder]: File not needed anymore.

File: logstash/config/.placeholder [hunk 7]
```diff
@@ -0,0 +1,5 @@
+Ensure the existence of the parent folder.
```
Comment (antoineco) [inline logstash/config/.placeholder]: File not needed anymore.

File: logstash/pipeline/logstash.conf [hunk 8]
```diff
@@ -1,7 +1,8 @@
 input {
-	tcp {
-		port => 5000
-	}
+    tcp {
+        port  => 5000
+        codec => json
```
Comment (antoineco) [inline logstash/pipeline/logstash.conf]: What if people send raw lines, with Filebeat for example?

**Comment by dm** [comment]
Please note I didn't amend anything; merely rebased and squashed; solving a couple of conflicts... 
I'll fix some of these; but I just needed to make this work for my own needs; feel free to bring this into a branch and continue from there...

File: README.md (hunk 2 above)
Comment (dm) [inline README.md]: I think he did this in error (if you notice the Create index section it's correct)

File: logstash/pipeline/logstash.conf (hunk 8 above)
Comment (dm) [inline logstash/pipeline/logstash.conf]: Not needed on my use; I believe I'd change this on own fork or so; no reason to make it more complicated in my point of view. 😄

**Comment by antoineco** [comment]
Sure I understand that, and appreciate you're sharing your work here. PRs must work for everyone without breaking the core function of the project in order to be mergeable though. I'm hoping we can use this as a base to come up with a nice solution.

**Comment by joaquincasares** [comment]
Nice, thanks for the rebase! And apologies for the radio silence. I figured the example was out there, via a PR, but I'd be revisiting whenever bumped by someone on my side or yours.

Let me know if/when you want some movement here and I'll put in a bit of time. Hopefully I get a request to do a bit of work here soon. :)

**Comment by dm** [comment]
I can clean that bit out and make it persistent on my side; either should be fine if it helps this move along too ;)

File: README.md [hunk 9]
```diff
@@ -67,9 +79,29 @@ By default, the stack exposes the following ports:
...
+When Kibana launches, it will not contain any out-of-the-box indexes. Instead,
+you can run this command to create a Logstash index:
 
-*WARNING*: If you're using *Docker Toolbox*, you must access it via the *docker-machine* IP address instead of *localhost*.
+```bash
+$ curl -XPUT http://elasticsearch:9200/.kibana/index-pattern/logstash-* -d \
+    '{"title" : "logstash-*"}'
+```
+
+This command will mark the Logspout index as the default Kibana index:
```
Comment (deviantony) [inline README.md]: Do you mean Logstash index?

File: README.md [hunk 10]
```diff
@@ -81,11 +113,27 @@ The Kibana default configuration is stored in `kibana/config/kibana.yml`.
 
 It is also possible to map the entire `config` directory instead of a single file.
 
-## How can I tune the Logstash configuration?
```
Comment (deviantony) [inline README.md]: We should probably keep that one or update all the other section titles.

File: README.md [hunk 11]
```diff
@@ -81,11 +113,27 @@ The Kibana default configuration is stored in `kibana/config/kibana.yml`.
...
+## How can I tune Logstash configuration?
```
Comment (deviantony) [inline README.md]: This whole section was updated via https://github.com/deviantony/docker-elk/commit/ac945e0c96129aeb273f2ebd79d4dbafc79dc15c

Why would you want to revert the changes?

File: logstash/pipeline/logstash.conf (hunk 8 above)
Comment (deviantony) [inline logstash/pipeline/logstash.conf]: Agree with @antoineco 

Default behaviour should not be altered.

File: README.md [hunk 12]
```diff
@@ -2,15 +2,18 @@
 
 [![Join the chat at https://gitter.im/deviantony/docker-elk](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/deviantony/docker-elk?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
 
-Run the latest version of the ELK (Elasticsearch, Logstash, Kibana) stack with Docker and Docker-compose.
+Run the latest version of the ELK (Elasticsearch, Logstash, Kibana) stack with Docker and Docker Compose.
+Also includes Logspout which pipes Docker Log output into the ELK stack.
```
Comment (deviantony) [inline README.md]: I do not think that Logspout should be part of the 'core' elements of the stack, IMO it should be part of available extensions.

Should we introduce a new **Extensions** section in the README with the extra information about what brings the extension and how to start it within the stack?

File: README.md [hunk 13]
```diff
@@ -2,15 +2,18 @@
...
 
 It will give you the ability to analyze any data set by using the searching/aggregation capabilities of Elasticsearch and the visualization power of Kibana.
+With Logspout, we're able to automatically send all console loggings normally collected by Docker Logs into ELK without any additional configuration.
```
Comment (deviantony) [inline README.md]: See my comment above.

**Comment by dm** [comment]
@joaquincasares would be great if you could pick up the comments; specially since the original code is yours anyway... I'm not claiming ownership! 😄

File: README.md (hunk 9 above)
Comment (joaquincasares) [inline README.md]: Good catch on the typo!

File: README.md (hunk 11 above)
Comment (joaquincasares) [inline README.md]: This may have gotten lost during the rebase?

**Comment by joaquincasares** [comment]
Thanks for all the work and comments!

@dm Yeah, feel free to finish patching this up and getting this into master. I merely wanted to provide a great starting point for those interested in a Docker-ELK stack that includes Logspout.

Sorry for the silence! Friday's schedule was a bit different followed by a busy weekend. :)

**Comment by antoineco** [comment]
I definitely want to merge this after the minor issues mentioned before have been addressed.

The only thing that really bothers me is the complexity of the logspout image. I'm not too familiar with it so I'm not in a good position to propose something better, but I believe we can tackle this later in some way.

File: logstash/pipeline/logstash.conf (hunk 8 above)
Comment (antoineco) [inline logstash/pipeline/logstash.conf]: Well, my comment about Filebeat was actually not accurate because `beats` requires a special input on the Logstash side. However instead of altering the default tcp input I'll mention the requirement inside a README file (eg. inside extras/logspout/README.md). How does it sound?

File: README.md (hunk 12 above)
Comment (antoineco) [inline README.md]: I agree with that, and place the assets inside an `extensions` directory at the root of the repo maybe?

File: logstash/pipeline/logstash.conf (hunk 8 above)
Comment (deviantony) [inline logstash/pipeline/logstash.conf]: Agree, what about `extensions/logspout/README.md` instead of `extra/logspout/README.md` as you mentioned in the comment below ?

Comment (antoineco) [inline logstash/pipeline/logstash.conf]: Indeed, something like (repeating content of another comment):
```
extensions
├── logspout
│   ├── Dockerfile
│   ├── README.md
│   ├── config
│   ├── logspout-compose.yml
│   └── modules.go
└── some-other-ext
    ├── Dockerfile
    ├── README.md
    └── some-other-ext-compose.yml
```

File: README.md (hunk 10 above)
Comment (antoineco) [inline README.md]: Fixed in latest commit.

**Comment by antoineco** [comment]
I pushed quite a lot of changes today in order to adopt the discussed directory structure and add the requested changes.

@deviantony @dm and @joaquincasares you should `git reset --hard` your local repo in order to pick up the changes.

I'd say the branch is clean enough to be merged now, let's wait for Anthony's approval.

**Comment by antoineco** [comment]
@deviantony yeah we definitely should, I just added a short paragraph about it ✍️

**Comment by deviantony** [comment]
Did you just created the `extensions/README.md` file? I was thinking about a section inside the main `README.md` listing available extensions (somehow equivalent to the branch features listed at the beginning of the README).

**Comment by antoineco** [comment]
We made it 👏  Thanks a lot everybody!

**Comment by dm** [comment]
Great, thanks for this all!
@antoineco @deviantony and @joaquincasares 

I'll bring these changes to my fork as well, but keep my work-specific changes on a separate branch there for the time being - hope to consolidate it in the future and keep some contributions going.

Cheers,
Dan :)
